packages/models/src/store/store.ts

```diff
@@ -72,22 +146,184 @@ export class Store<
       result = result?.[paths[i]]
     }
     if (result === undefined || result === null) {
-      throw new NonExistentException(paths.join('.'))
+      throw new NonExistentException(`${key}:${paths.join('.')}`)
     }
     return result
   }
 
-  handleCall = (_msg: ActorMessage<MessagePayload<StoreMessage<GetStorageStruct<StructSchema>>>>): void => {
+  private cloneScript<T extends 'lock' | 'type'>(scriptValue: unknown, type: T): GetFullStorageStruct<StructSchema>[T] {
+    if (typeof scriptValue !== 'object' || scriptValue === null) throw new Error()
+    const res = {} as GetFullStorageStruct<StructSchema>[T]
+    if ('args' in scriptValue) {
+      this.assetStorage(this.getStorage([type, 'args']))
+      res['args'] = this.getStorage([type, 'args'])?.clone(scriptValue.args)
+    }
+    if ('codeHash' in scriptValue) {
+      this.assetStorage(this.getStorage([type, 'codeHash']))
+      res['codeHash'] = this.getStorage([type, 'codeHash'])?.clone(scriptValue.codeHash)
+    }
+    return res
+  }
+
+  private updateChainData(key: OutPointString, paths?: StorePath) {
+    if (!paths) {
+      const value = this.get(key)
+      if ('data' in value) {
+        this.updateValueInCell('data', key, value.data)
+      }
+      if ('witness' in value) {
+        this.updateValueInCell('witness', key, value.witness)
+      }
+      if ('lock' in value) {
+        this.updateScriptInCell('lock', key, value.lock)
+      }
+      if ('type' in value) {
+        this.updateScriptInCell('type', key, value.type)
+      }
+    } else {
+      switch (paths[0]) {
+        case 'data':
+        case 'witness':
+          this.updateValueInCell(paths[0], key, this.get(key, [paths[0]]))
+          break
+        case 'lock':
+        case 'type':
+          this.updateScriptInCell(paths[0], key, this.get(key)[paths[0]])
+          break
+        default:
+          break
+      }
+    }
+  }
+
+  private updateScriptInCell(
+    type: 'lock' | 'type',
+    key: OutPointString,
+    newValue: GetFullStorageStruct<StructSchema>['lock'] | GetFullStorageStruct<StructSchema>['type'],
+  ) {
+    if ('args' in newValue) {
+      this.updateValueInCell([type, 'args'], key, newValue.args)
+    }
+    if ('codeHash' in newValue) {
+      this.updateValueInCell([type, 'codeHash'], key, newValue.codeHash)
+    }
```

Comment:
`hashType` should be configurable too, if `code hash` is writable.
At least they should be paired because `code hash + hash type` works together as a pointer to a script.

Reply:
There is an update for off-chain data, I thought `hashType` will not store off-chain data.

Reply:
I didn't get the point of `update for off-chain data`

Say that we have a `store` which is a group of cells with `lock script`
```
code_hash: A
hash_type: type
args: B
```
And a request is to change their lock algorithm to
```
code_hash: C
hash_type: data
args: B
```
How can we achieve that?

Reply:
I got that, so I need to add a function to update `code_hash` or `hash_type`.
This function is an updated side-effect by [set](https://github.com/yanguoyu/kuai/blob/aa6cbf53aae3f599c496865df1b76d18ea36a888/packages/models/src/store/store.ts#L420)

Reply:
But the `store` is designed to manage `schema data`, should we need to manage `cell` data?

Reply:
> But the `store` is designed to manage `schema data`, should we need to manage `cell` data?

`Store` is an abstraction layer to read and write data of cells. Data is a general concept, including every field of a cell, `type script`, `lock script`, `capacity`, `output data`, and `witness`.

Meanwhile, `type script` and `lock script` have their own schema, in the schema, `hash type` and `code hash` have pre-defined schema `hash type: type | data`, `code hash: string` and `args` has a flexible schema which should be defined by users.

Reply:
But actually, `hash_type` or `code_hash` are meaningless to `serialize` or `deserialize`.
So I defined schema like this
```
{
  data?: FieldSchema<T>
  witness?: FieldSchema<T>
  lockArgs?: FieldSchema<T>
  typeArgs?: FieldSchema<T>
}
```

Reply:
> But actually, `hash_type` or `code_hash` are meaningless to `serialize` or `deserialize`. So I defined schema like this
> 
> ```
> {
>   data?: FieldSchema<T>
>   witness?: FieldSchema<T>
>   lockArgs?: FieldSchema<T>
>   typeArgs?: FieldSchema<T>
> }
> ```

I think it's clear, `code hash` and `hash type` could be updated by `lockArgs.codeHash =` and `lockArgs.hashType =` while `args` could be updated by custom schema as `lockArgs.args.flag =`

Reply:
Yeah, `data`, `witness`, `lockArgs`, `typeArgs ` can be updated or saved by the `set` function or `update_cells` message. And the `set` function has a side effect on `cell`.
If they need to update `hash_type` or `code_hash`, I think I can add an update function to update these.

Reply:
> Yeah, `data`, `witness`, `lockArgs`, `typeArgs ` can be updated or saved by the `set` function or `update_cells` message. And the `set` function has a side effect on `cell`. If they need to update `hash_type` or `code_hash`, I think I can add an update function to update these.

Better to add an `update hash type` option because updating `code hash` has its own, back to the topic,
`hash type` should be accessible in `updateScriptInCell`(https://github.com/yanguoyu/kuai/blob/aa6cbf53aae3f599c496865df1b76d18ea36a888/packages/models/src/store/store.ts#L192-L203) because `code hash` requires it to be changed in some cases.

Reply:
https://github.com/yanguoyu/kuai/blob/aa6cbf53aae3f599c496865df1b76d18ea36a888/packages/models/src/store/store.ts#L192-L203 is called when developers call `set` to update `state`, so I need to sync these changes to `cell`.
But the `schema`s don't include `hash_type`, so when developers call `set` to update `state`, it doesn't need to update `hash_type`, I think I should expose a public function to update `hash_type` or `code_hash` only.
And after I update the schema definition to
```
{
  data?: FieldSchema<T>
  witness?: FieldSchema<T>
  lockArgs?: FieldSchema<T>
  typeArgs?: FieldSchema<T>
}
```
I have deleted the `updateScriptInCell` function.